Graph printing + Print from a case detail

amstone326 opened the pull request:
cross-request: https://github.com/dimagi/commcare-core/pull/527

amstone326 commented:
@damagatchi retest this please

ctsims commented on app/build.gradle:
```diff
@@ -191,6 +192,7 @@ android {
     defaultConfig {
         minSdkVersion 9
         targetSdkVersion 25
+        multiDexEnabled true
```
pretty big change to move to multiDex. This is 100% needed for tihs work, right?

amstone326 replied:
Yea, the app won't run without it. It seems like adding the jsoup library just took us over the limit. We could try to go through and see if there are any libraries we can remove, but unless we removed a lot, we'd probably just hit this again sometime soon.

ctsims commented on app/src/org/commcare/graph/model/GraphData.java:
```diff
@@ -63,4 +73,91 @@ public String getConfiguration(String key, String defaultValue) {
+                            "</html>");
+        } catch (JSONException e) {
+            e.printStackTrace();
+            throw new RuntimeException(e);
```
Shouldn't this be a GraphException?

amstone326 replied:
Not sure; I just copied this code over from one place to another, so this is how it already was, but that seems right to me so I'll change it.

ctsims commented on app/src/org/commcare/print/PrintableDetailField.java:
```diff
@@ -0,0 +1,100 @@
+            String fullGraphHtml = ((GraphData)evaluatedField).getGraphHTML(graphTitle);
+            this.valueString = fullGraphHtml;
+        } catch (GraphException e) {
+            this.valueString = "";
```
This seems like it's going to make this pretty hard to debug if there are any issues

ctsims commented on app/src/org/commcare/activities/EntityDetailActivity.java:
```diff
@@ -251,4 +263,42 @@ public void onBackPressed() {
+    private void printDetail() {
+        Intent i = new Intent(this, TemplatePrinterActivity.class);
+        i.putExtra(TemplatePrinterActivity.PRINT_TEMPLATE_REF_STRING,
+                detail.getDerivedPrintTemplatePath());
```
So for clarity in a tabbed detail view, only the "Parent" detail is currently capable of printing right now, right?

amstone326 replied:
Yep. mLabour was very clear that the only thing they were interested in was being able to print the entire detail at once, so I just went with implementing that for now.

ctsims replied:
cool

amstone326 commented:
@ctsims So I'm tempted to disagree about the global template thing. In the short term it actually has to stay in there because the HQ-side work for detail printing isn't done yet, so there's currently no other way to set the template there while we're testing. But there have also been a couple of times in semi-recent memory that that HQ document upload stopped working in some way or another, and it was really nice to have the global template to fall back on. But if you're definitely still opposed to it, I'll make a note to get rid of it once the HQ side is done and we've confirmed that that method of template upload is working consistently.

Other than that, addressed all of your comments here.

ctsims commented on app/src/org/commcare/print/PrintableDetailField.java:
```diff
@@ -0,0 +1,100 @@
+            String fullGraphHtml = ((GraphData)evaluatedField).getGraphHTML(graphTitle);
+            this.valueString = fullGraphHtml;
+        } catch (GraphException e) {
+            this.valueString = "<p>ERROR ENCOUNTERED WHILE TRYING TO PRINT GRAPH</p>";
```
We should put the error text in this message, right? So people can figure out what failed?
